scapy/layers/dns.py

```diff
@@ -19,12 +19,161 @@
+
+    def field_gen(dns_pkt):
+        """Iterates through all DNS strings that can be compressed"""
+        for lay in [dns_pkt.qd, dns_pkt.an, dns_pkt.ns, dns_pkt.ar]:
+            if lay is None:
+                continue
+            current = lay
+            while not isinstance(current, NoPayload):
+                if isinstance(current, InheritOriginDNSStrPacket):
+                    for field in current.fields_desc:
+                        if isinstance(field, DNSStrField) or \
+                           (isinstance(field, RDataField) and
+                           current.type in [2, 5, 12]):
+                                # Get the associated data and store it accordingly  # noqa: E501
+                                dat = current.getfieldval(field.name)
+                                yield current, field.name, dat
+                current = current.payload
+
+    def possible_shortens(dat):
+        """Iterates through all possible compression parts in a DNS string"""
+        yield dat
+        for x in range(1, dat.count(b".")):
+            yield dat.split(b".", x)[x]
+    data = {}
+    burned_data = 0
+    dummy_dns = DNSStrField("", "")  # Used for its i2m method
+    for current, name, dat in field_gen(dns_pkt):
+        for part in possible_shortens(dat):
+            # Encode the data
+            encoded = dummy_dns.i2m(None, part)
+            if part not in data:
+                # We have no occurrence of such data, let's store it as a
+                # possible pointer for future strings.
+                # We get the index of the encoded data
+                index = build_pkt.index(encoded)
+                index -= burned_data
+                # The following is used to build correctly the pointer
+                fb_index = ((index >> 8) | 0xc0)
+                sb_index = index - (256 * (fb_index - 0xc0))
+                pointer = chb(fb_index) + chb(sb_index)
+                data[part] = [(current, name, pointer)]
+            else:
+                # This string already exists, let's mark the current field
+                # with it, so that it gets compressed
+                data[part].append((current, name))
+                # calculate spared space
+                burned_data += len(encoded) - 2
+                break
+    # Apply compression rules
+    for ck in data:
+        # compression_key is a DNS string
+        replacements = data[ck]
+        # replacements is the list of all tuples (layer, field name)
+        # where this string was found
+        replace_pointer = replacements.pop(0)[2]
+        # replace_pointer is the packed pointer that should replace
+        # those strings. Note that pop remove it from the list
+        for rep in replacements:
+            # setfieldval edits the value of the field in the layer
+            val = rep[0].getfieldval(rep[1])
+            assert val.endswith(ck)
+            new_val = dummy_dns.i2m(None, val[:-len(ck)])[:-1] + \
```

Comment:
I think that it is easier to read if written:
tmp = val[:-len(ck)])[:-1] + replace_pointer
new_val = dummy_dns.i2m(None, tmp)

Reply:
Hmm not exactly the same thing (you’re calling the i2m on the replace_pointer)